### PR TITLE
Dev branch pull request from Ranjith Raj D

### PR DESCRIPTION
**Pull request by ranraj**
Hi,

Please find the list of available items in this pull request

UserService endpoint available for User and Group
Code formatted with default Idea support
Changes done as per the last call discussion
Created EntityProxy class to persist Entity with Simple type ( Different approach to avoid persisting Map)

### CODE REVIEW

**Comment by ranraj** [comment]
- Option get function handled with map and for comprehension 
- Future[Unit] were written to handle delete functionality
- Exception handled in the fetchNodeById function
- toGraph implicit function renamed

File: README.md [hunk 1]
```diff
@@ -1,2 +1,44 @@
-# activeGrid
-activeGrid in Scala
+# Learning akka for web application (ActiveGrid)
```
Comment (eklavya) [inline README.md]: We don't need to change this file.

File: src/main/scala/com/imaginea/Main.scala [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+ */
+object Main extends App {
+  val logger = Logger(LoggerFactory.getLogger(getClass.getName))
+
+  implicit val config = ConfigFactory.load
+  implicit val system = ActorSystem()
+  implicit val materializer = ActorMaterializer()
+  implicit val executionContext = system.dispatcher
+
+  implicit val neo4jRepository = Neo4jRepository
```
Comment (eklavya) [inline src/main/scala/com/imaginea/Main.scala]: Why do we need an implicit val for Neo4jRepository?

File: src/main/scala/com/imaginea/activegrid/core/discovery/models/Instance.scala [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  val ImageInfo: ImageInfo = null
+  val existingUsers: List[InstanceUser] = List.empty
+
+}
+
+case class StorageInfo(used: Double,total: Double) extends BaseEntity
+
+case class SSHAccessInfo(keyPairkey: KeyPairInfo,keyName: String,port : Option[Int] = None) extends BaseEntity{
+  def this(keyName: String){
+    this(new KeyPairInfo(None,keyName, null, null, null,KeyPairStatus.NOT_YET_UPLOADED,null,null),null)
```
Comment (eklavya) [inline src/main/scala/com/imaginea/activegrid/core/discovery/models/Instance.scala]: No nulls, whatsoever.

File: src/main/scala/com/imaginea/activegrid/core/models/Neo4jRep.scala [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * Created by babjik on 4/10/16.
+ */
+trait Neo4jRep[T] {
+  def toNeo4jGraph(entity: T): Option[Node]
+
+  def fromNeo4jGraph(nodeId: Long): T
+}
+
+trait Neo4jRep2[T] {
```
Comment (eklavya) [inline src/main/scala/com/imaginea/activegrid/core/models/Neo4jRep.scala]: Why is this needed?

File: src/main/scala/com/imaginea/activegrid/core/models/User.scala [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                , uniqueId: String
+                , publicKeys: List[KeyPairInfo]
+                , accountNonExpired: Boolean
+                , accountNonLocked: Boolean
+                , credentialsNonExpired: Boolean
+                , enabled: Boolean
+                , displayName: String)
+  extends BaseEntity
+
+case class UserProxy(
```
Comment (eklavya) [inline src/main/scala/com/imaginea/activegrid/core/models/User.scala]: No proxies, for now.

File: src/main/scala/com/imaginea/activegrid/core/models/UserGroup.scala [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+      logger.debug(s"UserGroupProxy Node saved into db - ${userGroupNode}")
+      //val map: Map[String, Any] = Map("name" -> userGroup.name)
+      //val node = Neo4jRepository.saveEntity[KeyPairInfo](ResourceACLProtocol.label, userGroup.id, map)
+
+      //map function is used to extract the option value
+      userGroup.users.map { users =>
+        logger.debug(s"UserGroupProxy has relation with Users $users")
+        //Iterating the users and linking to the UserGroup
+        for {user <- users
```
Comment (eklavya) [inline src/main/scala/com/imaginea/activegrid/core/models/UserGroup.scala]: use foreach

File: src/main/scala/com/imaginea/activegrid/core/utils/FileUtils.scala [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import java.io.{File, PrintWriter}
+
+/**
+ * Created by babjik on 5/10/16.
+ */
+object FileUtils {
+
+  def saveFile(file: String, content: String): Unit = {
+    val writer = new PrintWriter(new File(file), "UTF-8")
+    try {
```
Comment (eklavya) [inline src/main/scala/com/imaginea/activegrid/core/utils/FileUtils.scala:16]: Don't we want to log when things fail?

File: src/main/scala/com/imaginea/activegrid/core/services/UserService.scala [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+    }
+    }
+
+    //TODO: need to add code to write data to file
+
+    val keysList = user.fromNeo4jGraph(userId).map(_.publicKeys).getOrElse(List.empty)
+    Some(Page(0, keysList.size, keysList.size, keysList))
+  }
+
+  def deleteUser(userId: Long): Future[Option[String]] = Future {
```
Comment (eklavya) [inline src/main/scala/com/imaginea/activegrid/core/services/UserService.scala]: Why is Future[Option[]], still here?

**Review by eklavya — CHANGES_REQUESTED** [review]

File: src/main/scala/com/imaginea/activegrid/core/models/User.scala (hunk 5 above)
Comment (ranraj) [reply src/main/scala/com/imaginea/activegrid/core/models/User.scala]: I am working on this removing proxy. I will push this changes later.

File: src/main/scala/com/imaginea/activegrid/core/models/Neo4jRep.scala (hunk 4 above)
Comment (ranraj) [reply src/main/scala/com/imaginea/activegrid/core/models/Neo4jRep.scala]: People are using `def fromNeo4jGraph(nodeId: Long): T` returning T as return type
where as Neo4jRep2 returning Option[T]
`def fromNeo4jGraph(nodeId: Long): Option[T]`

If you add your comments on this return type I can merge both trait

Comment (eklavya) [reply src/main/scala/com/imaginea/activegrid/core/models/Neo4jRep.scala]: Can there be a case where the saved node isn't retrievable ?

File: src/main/scala/com/imaginea/activegrid/core/services/UserService.scala (hunk 8 above)
Comment (ranraj) [reply src/main/scala/com/imaginea/activegrid/core/services/UserService.scala]: Seems to be a Merge conflict; The latest Babji's code has not got merged with this.

My changes are available in the flowing userGroup function.

```
 def deleteUserGroup(userGroupId: Long): Future[Unit] =
    Future {
      Neo4jRepository.removeEntity[UserGroupProxy](userGroupId)
    }
```

File: README.md (hunk 1 above)
Comment (ranraj) [reply README.md]: Done

File: src/main/scala/com/imaginea/activegrid/core/discovery/models/Instance.scala (hunk 3 above)
Comment (ranraj) [reply src/main/scala/com/imaginea/activegrid/core/discovery/models/Instance.scala]: Done

File: src/main/scala/com/imaginea/activegrid/core/models/UserGroup.scala (hunk 6 above)
Comment (ranraj) [reply src/main/scala/com/imaginea/activegrid/core/models/UserGroup.scala]: Done

File: src/main/scala/com/imaginea/Main.scala (hunk 2 above)
Comment (ranraj) [reply src/main/scala/com/imaginea/Main.scala]: Done

File: src/main/scala/com/imaginea/activegrid/core/models/Neo4jRep.scala (hunk 4 above)
Comment (ranraj) [reply src/main/scala/com/imaginea/activegrid/core/models/Neo4jRep.scala]: Done

File: src/main/scala/com/imaginea/activegrid/core/models/User.scala (hunk 5 above)
Comment (ranraj) [reply src/main/scala/com/imaginea/activegrid/core/models/User.scala]: Done

File: src/main/scala/com/imaginea/activegrid/core/utils/FileUtils.scala (hunk 7 above)
Comment (ranraj) [reply src/main/scala/com/imaginea/activegrid/core/utils/FileUtils.scala:16]: Done

**Comment by ranraj** [comment]
The bellow issues were fixed and updated to the same PR.
- Enumeration has been moved to sealed trait
- Null values are removed
- Used Option instead of using Some("")
- Proxy case class replaced with Map of Entity property

and all other comments are fixied

File: src/main/scala/com/imaginea/activegrid/core/discovery/models/Filter.scala [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    case "SECURITY_GROUPS" => SecurityGroups
+    case "IP_RANGES" => IpRanges
+    case "INSTANCE_IDS" => InstanceIds
+    case "STATUS" => Status
+  }
+}
+
+sealed trait Condition {
+  def name: String
+  override def toString: String = name
```
Comment (eklavya) [inline src/main/scala/com/imaginea/activegrid/core/discovery/models/Filter.scala:38]: Why do we need this?

File: src/main/scala/com/imaginea/activegrid/core/models/ImageInfo.scala [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+object ImageInfo {
+
+  implicit class RichImageInfo(imageInfo: ImageInfo) extends Neo4jRep[ImageInfo] {
+    val logger = Logger(LoggerFactory.getLogger(getClass.getName))
+    val label = "ImageInfo"
+
+    override def toNeo4jGraph(entity: ImageInfo): Option[Node] = {
+      logger.debug(s"toGraph for Image ${imageInfo}")
+      // TODO: Image fields
+      val map: Map[String, Any] = Map()
```
Comment (eklavya) [inline src/main/scala/com/imaginea/activegrid/core/models/ImageInfo.scala]: Why Map of Any?

**Review by eklavya — CHANGES_REQUESTED** [review]

File: src/main/scala/com/imaginea/activegrid/core/discovery/models/Filter.scala (hunk 9 above)
Comment (ranraj) [reply src/main/scala/com/imaginea/activegrid/core/discovery/models/Filter.scala:38]: Condition is part of Filter case class. It was written as Enum in java and in scala we have moved it to the sealed Trait.

When we are persisting this Condition type that should be converted as String.
So I am holding name property and overriding it in the all other sub types.
I did refer this below link,
http://stackoverflow.com/questions/1898932/case-objects-vs-enumerations-in-scala

File: src/main/scala/com/imaginea/activegrid/core/models/ImageInfo.scala (hunk 10 above)
Comment (ranraj) [reply src/main/scala/com/imaginea/activegrid/core/models/ImageInfo.scala]: Neo4j accept only primitive values as the persistence type.
It could be AnyVal but String is in under AnyRef. So , we made it as Any.

As per your comments in other repository I have removed this type specification in UserGroup. 
This ImageInfo entity is not yet used in any of the endpoints. I forgot that change it.

**Comment by ranraj** [comment]
Option removed
Added exception handling in User Group and Access endpoints
Routing logic moved from Main to Controller

**Comment by ranraj** [comment]
- Exception logged as per the comment
- Validation done using forAll function
- Validation returning None if property is not available
- Removed unwanted code

**Comment by ranraj** [comment]
WIP : Rebasing with Master
